rts/motoko-rts/src/idl.rs

```diff
@@ -604,15 +613,48 @@ unsafe fn sub(
     // exit either via 'return true' or 'break 'return_false' to memoize the negative result
     'return_false: loop {
         match (u1, u2) {
-            (_, IDL_CON_alias) | (IDL_CON_alias, _) => idl_trap_with("sub: unexpected alias"),
-            (_, IDL_PRIM_reserved)
-            | (IDL_PRIM_empty, _)
-            | (IDL_PRIM_nat, IDL_PRIM_int)
-            | (_, IDL_CON_opt) => return true, // apparently, this is admissable
+            (IDL_CON_alias, IDL_CON_alias) => match mode {
+                CompatibilityMode::Candid => idl_trap_with("sub: unexpected alias"),
+                CompatibilityMode::Persistence => {
+                    let t11 = sleb128_decode(&mut tb1);
+                    let t21 = sleb128_decode(&mut tb2);
+                    // invariance
+                    if sub(rel, p, typtbl1, typtbl2, end1, end2, t11, t21, mode)
```

Comment:
I think there's a simpler and more efficient way to check invariant subtyping. Extend p to a three way enum (co, contra, inv) and, in the invariant cases, add the relation to both caches/and check the relation holds in both caches.
Then you only need on recursive call, not two (I expect the two calls probably lead to exponential behaviour).

Reply:
Thank you very much for this suggestion. Indeed, this would avoid potential exponential calls. I will implement this.